src/combinatorics/simple_graph/acyclic.lean

```diff
@@ -0,0 +1,137 @@
+  between pairs of vertices from a nonempty vertex type.
+
+## Tags
+
+acyclic graphs, trees
+-/
+
+universes u v
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/-- A graph is *acyclic* (or a *forest*) if it has no cycles. -/
+def is_acyclic : Prop := ∀ (v : V) (c : G.walk v v), ¬c.is_cycle
+
+/-- A *tree* is a connected acyclic graph. -/
+def is_tree : Prop := G.connected ∧ G.is_acyclic
+
+variables {G}
+
+lemma is_acyclic_iff_forall_adj_is_bridge :
+  G.is_acyclic ↔ ∀ {v w : V}, G.adj v w → G.is_bridge ⟦(v, w)⟧ :=
+begin
+  simp_rw [is_bridge_iff_adj_and_forall_cycle_not_mem],
+  split,
+  { intros ha v w hvw,
+    apply and.intro hvw,
+    intros u p hp,
+    exact absurd hp (ha _ p), },
+  { rintros hb v (_ | @⟨_, _, _, ha, p⟩) hp,
+    { exact hp.not_of_nil },
+    { specialize hb ha,
+      apply hb.2 _ hp,
+      rw [walk.edges_cons],
+      apply list.mem_cons_self } },
+end
+
+lemma is_acyclic_iff_forall_edge_is_bridge :
+  G.is_acyclic ↔ ∀ (e ∈ G.edge_set), G.is_bridge e :=
+by simp [is_acyclic_iff_forall_adj_is_bridge, sym2.forall]
+
+lemma is_acyclic.path_unique {G : simple_graph V} (h : G.is_acyclic) {v w : V} (p q : G.path v w) :
+  p = q :=
+begin
+  obtain ⟨p, hp⟩ := p,
+  obtain ⟨q, hq⟩ := q,
+  simp only,
+  induction p with u pu pv pw ph p ih generalizing q,
+  { cases q,
+    { refl },
+    { simpa [walk.is_path_def] using hq, } },
+  { rw is_acyclic_iff_forall_adj_is_bridge at h,
+    specialize h ph,
+    rw is_bridge_iff_adj_and_forall_walk_mem_edges at h,
+    replace h := h.2 (q.append p.reverse),
+    simp only [walk.edges_append, walk.edges_reverse, list.mem_append, list.mem_reverse] at h,
+    cases h,
+    { cases q,
+      { simpa [walk.is_path_def] using hp },
+      { rw walk.cons_is_path_iff at hp hq,
+        simp only [walk.edges_cons, list.mem_cons_iff, sym2.eq_iff] at h,
+        obtain (⟨h,rfl⟩ | ⟨rfl,rfl⟩) | h := h,
+        { rw [ih hp.1 _ hq.1] },
+        { simpa using hq },
+        { exact absurd (walk.fst_mem_support_of_mem_edges _ h) hq.2 } } },
+    { rw walk.cons_is_path_iff at hp,
+      exact absurd (walk.fst_mem_support_of_mem_edges _ h) hp.2 } }
+end
+
+lemma is_acyclic_of_path_unique (h : ∀ (v w : V) (p q : G.path v w), p = q) : G.is_acyclic :=
+begin
+  intros v c hc,
+  simp only [walk.is_cycle_def, ne.def] at hc,
+  cases c,
+  { exact absurd rfl hc.2.1 },
+  { simp only [walk.cons_is_trail_iff, not_false_iff, walk.support_cons,
+      list.tail_cons, true_and] at hc,
+    specialize h _ _ ⟨c_p, by simp only [walk.is_path_def, hc.2]⟩ (path.singleton (G.symm c_h)),
+    simp only [path.singleton] at h,
+    simpa [-quotient.eq, sym2.eq_swap, h] using hc },
+end
+
+lemma is_acyclic_iff : G.is_acyclic ↔ ∀ (v w : V) (p q : G.path v w), p = q :=
+⟨is_acyclic.path_unique, is_acyclic_of_path_unique⟩
+
+lemma is_tree_iff : G.is_tree ↔ nonempty V ∧ ∀ (v w : V), ∃!(p : G.walk v w), p.is_path :=
```

Comment:
Is using `∃!` really a good idea? It has little support.

Reply:
I think this lemma should stay as it is, but do we have something to get the path from v to w in a tree (or connected graph) in a convenient way?

Reply:
Not yet -- there's something about that in `walks_and_trees` for getting the unique path between two vertices in a tree, but I was going to deal with that in a follow-up PR.

Reply:
I think I'm missing something - isn't that an immediate consequence of this lemma?

Reply:
It takes writing the definition, giving it a name, writing some documentation, and giving some supporting API. The plan is to add these in a follow-up PR.